docs/proposals/pod-resource-management.md

```diff
@@ -0,0 +1,394 @@
+
+
+## Non Goals
+
+- Pod level disk accounting will not be tackled in this proposal.
+- Pod level resource specification in the Kubernetes API will not be tackled in this proposal.
+
+## Motivations
+
+Kubernetes currently supports container level isolation only and lets users specify resource requests/limits on the containers [Compute Resources](../../docs/design/resources.md). Kubelet creates cgroups sandbox (via docker) for all these containers.
+
+
+There are a few shortcomings to the current model.
+ - Existing Quality of Service support does not apply to pods as a whole. On-going work to support pod level eviction using QoS requires all containers in a pod to belong to the same class. By having pod level cgroups, it is easy to track pod level usage and make eviction decisions.
+ - Infrastructure overhead per pod is currently charged to the node. The overhead of running of setting up and managing the pod sandbox is currently accounted to the node. If the pod sandbox is a bit expensive, like in the case of hyper, having pod level accounting becomes critical.
+ - For the docker runtime we have a containerd-shim which is a small library that sits in front of a runtime implementation allowing it to be repartented to init, handle reattach from the caller etc. With pod level cgroups containerd-shim can be charged to the pod instead of the machine.
+ - If a container exits, all its anonymous pages (tmpfs) gets accounted to the machine (root). With pod level cgroups, that usage can also be attributed to the pod.
+ - Let containers share resources - with pod level limits, containers can choose to burst on demand and still be capped at their pod limits.
+
+## Design
+
+High level requirements for the design are as follows:
+ - Do not break existing users. Ideally, there should be no changes to the Kubernetes API semantics.
+ - Support multiple cgroup managers - systemd, cgroupfs, etc.
+
+How we intend to achieve these high level goals is covered in greater detail in the Implementation Plan.
+
+Examples in the next section.
+
+### Proposed cgroup hierarchy:
+
+For the initial implementation we will only support limits for cpu and memory resources.
+
+#### Quality of Service classes
+
+A pod can belong to one of the following 3 QoS classes: Guaranteed, Burstable, and Best-Effort, in decreasing order of priority.
+We use the following denotation:
+
+`G⇒ Guaranteed QoS, Bu⇒ Burstable QoS, BE⇒ Besteffort Qos`
+
+#### Guaranteed
+
+Guaranteed pods will be placed at the `$Root` cgroup by default. `$Root` is the system root ie. "/" by default and if `--cgroup-root` flag is used then we use the specified cgroup-root as the `$Root`.
+`Note: The cgroup-root flag would allow the user to configure the root of the QoS cgroup hierarchy. Hence cgroup-root would be redefined as the root of QoS cgroup hiearchy and not actually pods.`
+
+```
+/PodUID/cpu.quota = cpu limit of Pod  
```

Comment:
This is not clarified, but does this mean if I have a pod with UID: `5f9b19c9-3a30-11e6-8eea-28d2444e470d`

which cgroup path do I get:

1. /5f9b19c9-3a30-11e6-8eea-28d2444e470d/cpu.quota
2. /pod5f9b19c9-3a30-11e6-8eea-28d2444e470d/cpu.quota
3. /pod-5f9b19c93a3011e6-8eea28d2444e470d/cpu.quota

I am hoping 3 so I can inspect /sys/fs/cgroup and know this is a pod...